src/python/pants/backend/go/goals/package_binary_integration_test.py

```diff
@@ -102,6 +102,30 @@ def test_package_simple(rule_runner: RuleRunner) -> None:
     assert result.stdout == b"Hello world!\n"
 
 
+def test_package_third_party_requires_main(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "go.mod": dedent(
+                """\
+                module foo.example.com
+                go 1.17
+
+                require github.com/tgolsson/example-pants-third-party v0.0.0-20221101220057-1a7167a87ec5 // indirect
+                """
+            ),
+            "BUILD": dedent(
+                """\
+                go_mod(name='mod')
+                go_binary(name="bin", main='//:mod#github.com/tgolsson/example-pants-third-party/pkg/hello')
+                """
+            ),
+        }
+    )
+    binary_tgt = rule_runner.get_target(Address("", target_name="bin"))
+    with engine_error(ValueError, contains="but uses package name `hello` instead of `main`"):
+        build_package(rule_runner, binary_tgt)
```

Comment:
It'd probably be good to have a positive case, too, where we don't error.

Reply:
It very closely mirrors the run_binary test but added here too as it's tests the file from outside pants.